assembler.py

```diff
@@ -55,15 +55,43 @@ def read_source(self, filename=None):
                 f'Unsupported file type [{self.filename}]. Only accepting files ending in .asm')
         source = open(self.filename, 'r')
         lines = source.readlines()
-        for line in lines:
-            if line != '\n':
-                self.micro_instr.append(line.strip())
+        if self.is_indented(lines[0]):
+            raise AssertionError('Indentation Error: the first line cannot be indented.')
+        else:
+            self.verify_indentation(lines[0], 0)
+            self.micro_instr.append(lines[0].strip())
+        for i in range(1, len(lines)):
+            self.verify_indentation(lines[i], i)
+            self.compare_indentation_between_lines(lines[i - 1], lines[i], i)
+            if lines[i] != '\n':
+                self.micro_instr.append(lines[i].strip())
         lines.clear()
         source.close()
 
+    def verify_indentation(self, line, index):
+        if "\t" in line:
+            raise AssertionError(f'Indentation error: Line {index + 1}: Tab detected.')
+        if not self.is_indented(line) and line.startswith(" ") and not line.isspace():
+            raise AssertionError(f'Indentation error: Line {index + 1}: Ensure that '
+                                 f'all indented lines have exactly 4 spaces.')
+        if ":" in line and self.is_indented(line):
+            raise AssertionError(
+                f'Indentation error: Line {index + 1}: Lines with \':\' cannot be indented.')
+
+    def compare_indentation_between_lines(self, line1, line2, index):
+        if self.is_indented(line2) and ((not self.is_indented(line1) and ":" not in line1)
+                                        or line1.isspace() or line1 == '\n'):
+            raise AssertionError(f'Indentation Error: Verify lines {index} and {index + 1}')
+        if not self.is_indented(line2) and ":" in line1:
+            raise AssertionError(f'Indentation Error: Line {index + 1}: lines under label must be indented.')
+        return True
```

Comment:
Who uses this boolean?

Reply:
In an outdated implementation, it used to be necessary; I'll remove it in the next commit